### PR TITLE
Implement ripemd160 and hmac_sha512 RPCs

### PR DESCRIPTION
**Pull request by nchashch**
Closes #62

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -50,6 +50,8 @@ tower-http = { version = "0.6.1", features = ["trace"] }
 tracing = "0.1.40"
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
 zeromq = "0.4.0"
+ripemd = "0.1.3"
```
Comment (Ash-L2L) [inline Cargo.toml:53]: Both of these are available already in the `bitcoin` crate, see
https://docs.rs/bitcoin_hashes/0.14.0/bitcoin_hashes/hmac/index.html
https://docs.rs/bitcoin_hashes/0.14.0/bitcoin_hashes/ripemd160/index.html

File: src/server.rs [hunk 2]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
         ))
     }
+
+    async fn ripemd160(
+        &self,
+        request: tonic::Request<Ripemd160Request>,
+    ) -> std::result::Result<tonic::Response<Ripemd160Response>, tonic::Status> {
+        let message = request.into_inner().message;
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<GetBmmHStarCommitmentRequest>("message"))?
```
Comment (Ash-L2L) [inline src/server.rs:700]: This is a mistake, you should use `missing_field::<Ripemd160Request>("message")`

File: src/server.rs [hunk 3]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
             "not implemented",
         ))
     }
+
+    async fn ripemd160(
+        &self,
+        request: tonic::Request<Ripemd160Request>,
+    ) -> std::result::Result<tonic::Response<Ripemd160Response>, tonic::Status> {
+        let message = request.into_inner().message;
```
Comment (Ash-L2L) [inline src/server.rs:698]: You should destructure here to enforce handling any new fields, if the request message definition changes

File: src/server.rs [hunk 4]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
+    }
+
+    async fn hmac_sha512(
+        &self,
+        request: tonic::Request<HmacSha512Request>,
+    ) -> std::result::Result<tonic::Response<HmacSha512Response>, tonic::Status> {
+        let HmacSha512Request { secret, message } = request.into_inner();
+        let secret: Vec<u8> = secret
+            .ok_or_else(|| missing_field::<GetBmmHStarCommitmentRequest>("secret"))?
+            .decode_tonic::<GetBlockHeaderInfoRequest, _>("secret")?;
```
Comment (Ash-L2L) [inline src/server.rs:718]: This is a mistake, you should use `decode_tonic::<HmacSha512Request>("secret")`

File: src/server.rs [hunk 5]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<GetBmmHStarCommitmentRequest>("message"))?
```
Comment (Ash-L2L) [inline src/server.rs:720]: This is a mistake, you should use `missing_field::<HmacSha512Request>("message")`

File: src/server.rs [hunk 6]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
+            .decode_tonic::<GetBlockHeaderInfoRequest, _>("message")?;
```
Comment (Ash-L2L) [inline src/server.rs:721]: This is a mistake, you should use `decode_tonic::<HmacSha512Request>("message")`

File: src/server.rs [hunk 7]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
+    ) -> std::result::Result<tonic::Response<Ripemd160Response>, tonic::Status> {
+        let message = request.into_inner().message;
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<GetBmmHStarCommitmentRequest>("message"))?
+            .decode_tonic::<GetBlockHeaderInfoRequest, _>("message")?;
+        let mut hasher = ripemd::Ripemd160::new();
+        hasher.update(&message);
+        let hash = hasher.finalize();
+        let response = Ripemd160Response {
+            hash: Some(ConsensusHex::encode(&hash.as_slice().to_vec())),
```
Comment (Ash-L2L) [inline src/server.rs:706]: Hashes are fixed size, so you should convert to an array, not a vec/slice, which will encode the length prefix.
use `hash.to_byte_array()`

File: src/server.rs [hunk 8]
```diff
@@ -684,4 +690,45 @@ impl WalletService for Arc<crate::wallet::Wallet> {
...
+            .decode_tonic::<GetBlockHeaderInfoRequest, _>("secret")?;
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<GetBmmHStarCommitmentRequest>("message"))?
+            .decode_tonic::<GetBlockHeaderInfoRequest, _>("message")?;
+
+        let mut mac = HmacSha512::new_from_slice(&secret).unwrap();
+        mac.update(&message);
+        let hash = mac.finalize().into_bytes();
+        let response = HmacSha512Response {
+            hash: Some(ConsensusHex::encode(&hash.as_slice().to_vec())),
```
Comment (Ash-L2L) [inline src/server.rs:727]: Hashes are fixed size, so you should convert to an array, not a vec/slice, which will encode the length prefix.
use `hash.to_byte_array()`

**Review by Ash-L2L — CHANGES_REQUESTED** [review]

File: src/server.rs (hunk 7 above)
Comment (nchashch) [reply src/server.rs:706]: There is no `to_byte_array()` in `GenericArray`:

https://docs.rs/generic-array/0.14.7/generic_array/struct.GenericArray.html

Comment (Ash-L2L) [reply src/server.rs:706]: see https://github.com/LayerTwo-Labs/bip300301_enforcer/pull/63/files#r1815525170, this should be done via the `bitcoin` crate

File: src/server.rs (hunk 8 above)
Comment (Ash-L2L) [reply src/server.rs:726]: see https://github.com/LayerTwo-Labs/bip300301_enforcer/pull/63/files#r1815525170, this should be done via the bitcoin crate

File: Cargo.toml (hunk 1 above)
Comment (nchashch) [reply Cargo.toml:53]: `ripemd160` crate is depricated, they are suggesting to use `ripemd` instead: http://docs.rs/ripemd160

I think it is reasonable to add `hmac` as a separate dependency as well, there isn't any significant cost to doing it.

Comment (Ash-L2L) [reply Cargo.toml:53]: `bitcoin_hashes` does not depend on the `ripemd160` crate
https://crates.io/crates/bitcoin_hashes/0.14.0/dependencies

**Review by Ash-L2L — APPROVED** [review]

**Pull request merged**